scheds/rust/scx_layered/src/bpf/main.bpf.c

```diff
@@ -788,16 +790,112 @@ void BPF_STRUCT_OPS(layered_enqueue, struct task_struct *p, u64 enq_flags)
 			return;
 	}
 
-	bpf_for(idx, 0, nr_possible_cpus) {
-		s32 cand = (preempt_cursor + idx) % nr_possible_cpus;
+	if (!disable_topology) {
+		if (!(cachec = lookup_cache_ctx(cctx->cache_idx)) ||
+		    !(nodec = lookup_node_ctx(cctx->node_idx)))
+			return;
 
-		if (try_preempt(cand, p, cctx, tctx, layer, false)) {
-			/*
-			 * Round-robining doesn't have to be strict. Let's
-			 * not bother with atomic ops on $preempt_cursor.
-			 */
-			preempt_cursor = (cand + 1) % nr_possible_cpus;
+		struct bpf_cpumask *attempted = bpf_cpumask_create();
+		if (!attempted)
 			return;
+
+		struct bpf_cpumask *topo_cpus = bpf_cpumask_create();
+		if (!topo_cpus) {
+			bpf_cpumask_release(attempted);
+			return;
+		}
+
+		if (!cachec->cpumask) {
+			bpf_cpumask_release(attempted);
+			bpf_cpumask_release(topo_cpus);
+			return;
+		}
+
+		bpf_cpumask_copy(topo_cpus, (const struct cpumask*)cachec->cpumask);
+
+		/*
+		 * First try preempting in the local LLC
+		 */
+		bpf_for(idx, 0, cachec->nr_cpus) {
+			s32 preempt_cpu = bpf_cpumask_any_distribute((const struct cpumask*)topo_cpus);
```

Comment:
Does this reliably try out the CPUs when looped upon?

Reply:
Added in a trace and it seems to work.

Reply:
@htejun I check this again and the kfunc is calling [`cpumask_first`](https://git.kernel.org/pub/scm/linux/kernel/git/torvalds/linux.git/tree/include/linux/cpumask.h#n253). Probably need to fix this to make it more fair.